ddtrace/contrib/kombu/patch.py

```diff
@@ -0,0 +1,91 @@
+# 3p
+import kombu
+import wrapt
+
+# project
+from ...pin import Pin
+from ...ext import kombu as kombux
+from ...utils.wrappers import unwrap
+from ...propagation.http import HTTPPropagator
+from .util import (
+    get_exchange_from_args,
+    get_body_length_from_args,
+    get_routing_key_from_args,
+    _extract_conn_tags
+)
+
+HEADER_POS = 4
+propagator = HTTPPropagator()
+
+
+def patch():
+    """Patch the instrumented methods
+
+    This duplicated doesn't look nice. The nicer alternative is to use an ObjectProxy on top
+    of Kombu. However, it means that any "import kombu.Connection" won't be instrumented.
+    """
+    if getattr(kombu, '_datadog_patch', False):
+        return
+    setattr(kombu, '_datadog_patch', True)
+
+    _w = wrapt.wrap_function_wrapper
+    _w('kombu', 'Producer._publish', traced_publish)
+    _w('kombu', 'Consumer.receive', traced_receive)
+    Pin(service="kombu", app="kombu", app_type="custom").onto(kombu.messaging.Producer)
+    Pin(service="kombu", app="kombu", app_type="custom").onto(kombu.messaging.Consumer)
+
+
+def unpatch():
+    if getattr(kombu, '_datadog_patch', False):
+        setattr(kombu, '_datadog_patch', False)
+        unwrap(kombu.Producer, '_publish')
+        unwrap(kombu.Consumer, 'receive')
+
+#
+# tracing functions
+#
+
+
+def traced_receive(func, instance, args, kwargs):
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    # Signature only takes 2 args: (body, message)
+    message = args[1]
+    context = propagator.extract(message.headers)
+    # only need to active the new context if something was propagated
+    if context.trace_id:
+        pin.tracer.context_provider.activate(context)
+    with pin.tracer.trace('kombu.Consumer.receive', service=pin.service, span_type='kombu') as s:
+        # run the command
+        exchange = message.delivery_info['exchange']
+        s.resource = exchange
```

Comment:
I'm not very familiar with Kombu, but what could be the cardinality of that? each single resource is computed as a metric, and if the cardinality is too high, the resource page results in something not really usable.

Reply:
We have (what I consider) a very large redis cluster, pushing thousands of messages per second, and have about 250 exchanges. I'd be more worried if we were doing this with queues which can fan-out and bind to exchanges, but as it is, it should be a stable number, depending on the app. They don't typically have UUIDs or some other high cardinality item in their name.

Reply:
According to the [documentation](http://docs.celeryproject.org/projects/kombu/en/latest/reference/kombu.html#exchange) it seems like exchanges should be appropriate to use for the resource.